rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,516 @@
+parameters.  Counters for these histograms can be sent to Couchbase and both
+stored individually and aggregated. See histogram buckets in "Histograms" table
+below.
+
+As an aggregated set of metrics, it would give Couchbase a rough view over the
+service level the user is experiencing.  Deviations in service level can be dug
+into and, in particular, we can check for correlation to Couchbase Capella
+management activities like upgrades.
+
+The Indexing Service is not included as it is not used by SDK, even for
+management operations. All future services must be reported if the service
+exposes separate port and the SDK is using it.
+
+As with the Service Experience Counters, it may be incorrect to try to
+compare between deployments.  For example, the query complexity may differ
+significantly or the data size and use of features like compression may vary
+significantly between two applications.
+
+The SDK must record latency for histogram on the lower layer for each
+transmission attempt instead of top-level operation. For example, if KV GET
+operation results in three network requests, there should be three data points
+for the metrics counter, and if the operation has not been canceled, it should
+record three data points for histogram, even if the operation has been timed
+out, because on the lower-level API, the SDK still sees orphan responses and is
+able to figure out true latency of the request.
+
+# API and Implementation Details
+
+The decision to reuse the existing Meter implementation is considered an SDK
+implementation detail, and this RFC does not specify the API for collecting
+metrics from the library.
+
+One possible approach is to introduce a "compound meter" that aggregates an
+existing meter (which could be overridden by the user) and the application
+service-level telemetry meter described in this RFC. This compound meter would
+collect counters and fixed-bucket latency histograms. The application
+service-level meter would continue to function even if the user provides a
+custom meter implementation.
+
+In addition to latencies, which would be used for reporting histograms, the
+default Meter should also maintain the following counters:
+
+| Metric | Description |
+| :---- | :---- |
+| sdk\_*{service}*\_r\_timedout | Timeout |
+| sdk\_*{service}*\_r\_canceled | Canceled |
+| sdk\_*{service}*\_r\_total | Total number of operations |
+
+The SDK must send zero metric, because it still carries information that at
+given timestamp, no events have been registered.
+
+Where the service is a part of the visible counter name, and hostname and
+bucket must be stored and exported as a meta data. Valid service names are (see
+RFC-67 for `Service Identifier`):
+
+* `kv`
+* `query`
+* `search`
+* `analytics`
+* `management`
+* `eventing`
+
+The SDK must only account network operations that are visible to server. In
+other words it must not increment counter for virtual/compound operations like
```

Comment:
I guess it's implied here, but could you spell out how transactions work - presumably it will appear as a set of unrelated subdoc reads and writes?

Reply:
Sure, I can update the spec explicitly say that.